File: Python/114. Flatten Binary Tree to Linked List.py
# 114. Flatten Binary Tree to Linked List

# Topic: Linked List, Stack, Tree, Depth-First Search, Binary Tree.

'''
Task: 
-----
Given the root of a binary tree, flatten the tree into a "linked list":
The "linked list" should use the same TreeNode class where the right child pointer points to the next node in 
the list and the left child pointer is always null.
The "linked list" should be in the same order as a pre-order traversal of the binary tree.

Example 1:
Input: root = [1,2,5,3,4,null,6]
Output: [1,null,2,null,3,null,4,null,5,null,6]

Example 2:
Input: root = []
Output: []

Example 3:
Input: root = [0]
Output: [0]
 
Constraints:
The number of nodes in the tree is in the range [0, 2000].
-100 <= Node.val <= 100
 
Follow up: Can you flatten the tree in-place (with O(1) extra space)?

# Testcase:
[1,2,5,3,4,null,6]
[]
[0]


# Code:
--------
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """

'''
# Solution:  ------------------------------------------------------------------

class TreeNode:
    """Definition for a binary tree node"""

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """Do not return anything, modify root in-place instead."""
        
        current = root
        while current:
            if current.left:
                # Find the rightmost node in the left subtree
                rightmost = current.left
                while rightmost.right:
                    rightmost = rightmost.right

                # Re-wire the connections
                rightmost.right = current.right
                current.right = current.left
                current.left = None

            # Move to the next node
            current = current.right

# Description:
"""
The solution provided for flattening a binary tree into a linked list in place follows a different approach than the recursive solution previously discussed. This solution iteratively modifies the tree by re-wiring the left subtree's rightmost node to the right subtree, ensuring the tree is flattened in a preorder traversal manner. Here's a step-by-step description of how this algorithm works:

### Algorithm Description:

1. **Initialization**: Start with the `current` node set to the root of the tree. The algorithm iterates through the tree starting from the root, moving towards the rightmost node.

2. **Iterating through the Tree**: The algorithm uses a `while` loop to traverse the tree. For each `current` node being processed, it checks if the `current` node has a left child.

3. **Processing the Left Subtree**:
    - If the `current` node has a left child, the algorithm finds the rightmost node in the left subtree. This is done by iterating through the left subtree using a `while` loop until the rightmost node is found.
    - Once the rightmost node in the left subtree is identified, its right pointer is set to the `current` node's right child. This effectively appends the entire right subtree to the end of the left subtree.

4. **Re-wiring Connections**:
    - The `current` node's right pointer is then updated to point to its left child, effectively moving the left subtree to the right side. This maintains the preorder sequence (root, left, right) since the left subtree is now positioned in the place of the right subtree.
    - The `current` node's left pointer is set to `None`, removing the original left subtree.

5. **Moving to the Next Node**: After re-wiring the connections for the `current` node, the algorithm moves to the next node by setting `current` to `current.right`. This step progresses the flattening process to the next node in the preorder sequence.

6. **Continuation Condition**: The loop continues until there are no more nodes to process, i.e., until `current` becomes `None`. At this point, the entire tree has been flattened into a linked list.

7. **In-Place Modification**: Throughout the process, the tree is modified in place. No new nodes are created, and the existing nodes are merely re-wired to achieve the flattened structure.

8. **Termination**: When the `while` loop exits, the tree has been successfully flattened into a linked list following the preorder traversal order, with all nodes' left pointers set to `None` and right pointers leading to the subsequent nodes in the list.

This iterative approach efficiently flattens the binary tree by carefully re-wiring the nodes' pointers, ensuring that the space complexity is kept to O(1) as no additional data structures are used and all modifications are performed in place.
"""
            

File: Cpp/114. Flatten Binary Tree to Linked List.cpp
// 114. Flatten Binary Tree to Linked List.

// Topic: Linked List, Stack, Tree, Depth-First Search, Binary Tree.

/*
Task: 
-----
Given the root of a binary tree, flatten the tree into a "linked list":
The "linked list" should use the same TreeNode class where the right child pointer points to the next node in 
the list and the left child pointer is always null.
The "linked list" should be in the same order as a pre-order traversal of the binary tree.

Example 1:
Input: root = [1,2,5,3,4,null,6]
Output: [1,null,2,null,3,null,4,null,5,null,6]

Example 2:
Input: root = []
Output: []

Example 3:
Input: root = [0]
Output: [0]
 
Constraints:
The number of nodes in the tree is in the range [0, 2000].
-100 <= Node.val <= 100
 
Follow up: Can you flatten the tree in-place (with O(1) extra space)?

# Testcase:
[1,2,5,3,4,null,6]
[]
[0]


# Code:
--------
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 *//*
class Solution {
public:
    void flatten(TreeNode* root) {
        
    }
};

*/

// Solution: ------------------------------------------------------------

class Solution {
public:
    void flatten(TreeNode* root) {
        if (!root) return; // Base case: if the tree is empty or we've reached a leaf node

        TreeNode* left = root->left;
        TreeNode* right = root->right;

        // Flatten the left subtree
        flatten(left);
        // Flatten the right subtree
        flatten(right);

        // Set the root's left to null and right to the previously left subtree
        root->left = nullptr;
        root->right = left;

        // Find the last node of the new right subtree
        TreeNode* current = root;
        while (current->right) {
            current = current->right;
        }

        // Attach the previously right subtree to the end of the new right subtree
        current->right = right;
    }
};


// Description:  =========================================================
/*
To flatten a binary tree into a linked list in the same order as a pre-order traversal, we can follow a recursive 
approach that adheres to the in-place requirement with O(1) extra space. The key idea is to traverse the tree in 
a preorder fashion (root, left, right) and rearrange the nodes such that each node's right child points to the next 
node in the preorder traversal, and the left child is always null.


### Solution Explanation:

1. **Recursive Approach**: The function `flatten` is called recursively for each node starting from the root. The recursion 
     flattens the left and right subtrees first before rearranging the nodes to form a linked list.

2. **Flattening Subtrees**: The left and right subtrees of the current node are flattened by recursive calls. After these 
     calls, both subtrees are flattened into linked lists themselves.

3. **Rearranging Nodes**: After flattening the left subtree, the root's left pointer is set to null, and its right pointer 
     is set to the head of the flattened left subtree. This effectively moves the flattened left subtree to the right, 
     maintaining the preorder sequence.

4. **Merging Subtrees**: The algorithm then iterates to the end of the newly formed right subtree (which was originally the 
    left subtree) and attaches the flattened right subtree to the end. This completes the flattening for the current node, 
    ensuring that the preorder traversal order is preserved.

5. **In-Place Transformation**: The tree is transformed in place; no new nodes are created, and the existing pointers are 
     merely rearranged to achieve the flattened structure.

6. **Base Case and Edge Cases**: The recursion has a base case when the current node is null, at which point the function 
     returns. This handles edge cases like an empty tree or reaching the end of a branch.

This solution ensures that the binary tree is flattened into a linked list following the preorder traversal order, with each 
node's right pointer leading to the next node in the list and the left pointer always set to null, achieving the desired 
transformation in place.

*/